python/pyspark/sql/datasource.py

```diff
@@ -298,6 +305,90 @@ def read(self, partition: InputPartition) -> Iterator[Union[Tuple, Row]]:
         ...
 
 
+class DataSourceStreamReader(ABC):
+    def initial_offset(self) -> dict:
+        """
+        The initial offset of the streaming data source.
+
+        Returns
+        -------
+        dict
+            A dict whose key and values are str type.
+        """
+        ...
+
+    def latest_offset(self) -> dict:
+        """
+        Seek the latest offset of the streaming data source.
+
+        Returns
+        -------
+        dict
+            A dict whose key and values are str type.
+        """
+        ...
+
+    def partitions(self, start: dict, end: dict) -> Sequence[InputPartition]:
+        """
+        Plan the read partitions given the start and end offset.
+
+        Parameters
+        ----------
+        start : dict
+            The start offset of the microbatch to plan partitioning.
+        end : dict
+            The end offset of the microbatch to plan partitioning.
+
+        Returns
+        -------
+        Sequence[InputPartition]
```

Comment:
It shouldn't actually use plain python type hints here but a description of the types (see the example in https://numpydoc.readthedocs.io/en/latest/format.html#parameters). It should be something like ``` sequence of :class:`InputPartition` ```

Reply:
I see there are same mistakes in the batch version so I don't mind fixing it separately for all later.

Reply:
Thanks, I feel it is probably cleaner to make separate change in another PR.